### PR TITLE
[WIP] Rand recomputation

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88217
* #87641

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Still need to add test. Fix for https://github.com/pytorch/pytorch/issues/93578

No cudagraphs, resnet18, before this pr:

cuda train resnet18                           memory: eager: 0.55 GB, dynamo: 0.79 GB, ratio: 0.70
1.082x p=0.00

No cudagraphs, resnet18, with this pr.
cuda train resnet18                           memory: eager: 0.55 GB, dynamo: 0.71 GB, ratio: 0.78
1.075x p=0.00

The memory savings are significant, but it does also introduce a slowdown. Thoughts ?

### CODE REVIEW

**Comment by Chillee** [comment]
I believe the slowdowns come from randomness being slow in Triton, perhaps this is something we could fix in Triton? cc: @ngimel @shintaro-iwasaki

**Comment by ngimel** [comment]
fx graph of resnet doesn't have dropout/rand, how can this affect it at all? Do you use some non-default settings?

**Comment by eellison** [comment]
@ngimel, hmm maybe, I’ll re-benchmark.

**Review by jansel — APPROVED** [review]
